documented comboBox (#443)
gsoft-inc/wl-hopper: apps/docs/content/components/pickers/ComboBox.mdx

---
title: Combobox
description: A combo box allows the user to make a selection from a list of suggested, likely or desired values.
category: "pickers"
links:
    source: https://github.com/gsoft-inc/wl-hopper/blob/main/packages/components/src/checkbox/src/Checkbox.tsx
    aria: https://www.w3.org/WAI/ARIA/apg/patterns/checkbox/
---

<Example src="ComboBox/docs/preview" isOpen />

## Props

<PropTable component="ComboBox" />

## Anatomy

<FeatureFlag flag="next">
    ## Guidelines

    TODO: If we have some guidelines about this component&apos;s usage

    ### Accessibility?

    TODO: If we have some guidelines about this component and accessibility
</FeatureFlag>

<FeatureFlag flag="rc">

    TODO: We have anatomy screenshots from the Figma, we could most likely use them here

    ### Concepts

    TODO: links to related concepts
</FeatureFlag>

### Composed Components

#### ComboBox

- [Label](./Label)
- [Helper Message](./HelperMessage)
- [Error Message](./ErrorMessage)

> - [Header](./Select#header)

Reviewer: ```suggestion
- [Header](./ComboBox#header)
```

> - [Section](./Select#section)

Reviewer: ```suggestion
- [Section](./ComboBox#section)
```

#### ComboBox.Option

- [Text](./Text)
- [Avatar](Avatar)
- [Icon](./Icon)
- [Badge](./Badge)

## ComboBox Examples

### Disabled
A combo box in a disabled state shows that it exists, but is not available in that circumstance. This can be used to maintain layout continuity and communicate that a field may become available later.

<Example src="ComboBox/docs/disabled" />

### Error
A combo box can be displayed in an error state to indicate that the selection is invalid.

<Example src="ComboBox/docs/error" />

### Sizes
A combo box has multiple sizes to choose from.
The combo box option also changes size based on the size of the combo box.

<Example src="ComboBox/docs/sizes" />

### Labeling

> If a visible label isn't specified, an `aria-label` must be provided to the select for accessibility. If the field is labeled by a separate element, an `aria-labelledby` prop must be provided using the ID of the labeling element instead.

Reviewer: ```suggestion
If a visible label isn't specified, an `aria-label` must be provided to the combo box for accessibility. If the field is labeled by a separate element, an `aria-labelledby` prop must be provided using the ID of the labeling element instead.
```

<Example src="ComboBox/docs/labeling" />

### Icon Prefix
An icon can be displayed at the start of the select trigger.

<Example src="ComboBox/docs/iconPrefix" />

### Text Prefix
A short text can be displayed at the start of the select trigger.

<Example src="ComboBox/docs/textPrefix" />

### Fluid
A combo box can take the width of its container.

<Example src="ComboBox/docs/fluid" />

### Controlled
A combo box can have a controlled selected value. In this example, it shows how it is possible to deselect an option.

<Example src="ComboBox/docs/controlled" />

### Custom Filtering

> A combo box can overridde the default [useFilter](https://react-spectrum.adobe.com/react-aria/useFilter.html) function from RAC. This can be overridden using the `defaultFilter` prop, or by using the items prop to control the filtered list. When items is provided rather than `defaultItems`, ComboBox does no filtering of its own.

Reviewer: ```suggestion
A combo box can overridde the default [useFilter](https://react-spectrum.adobe.com/react-aria/useFilter.html) function from RAC. This can be overridden using the `defaultFilter` prop, or by using the `items` prop to control the filtered list. When `items` is provided rather than `defaultItems`, ComboBox does no filtering of its own.
```

### Disabled Item

A combo box with a disabled item.

<Example src="ComboBox/docs/ComboBoxOption/disabled" />

### Auto Menu Width

A combo box can have a menu that automatically adjusts its width based on the longest item.

<Example src="ComboBox/docs/ComboBoxOption/autoMenuWidth" />

### Menu placement

A combo box's menu can have a customized menu placement using the `placement` prop inside `popoverProps`.

<Example src="ComboBox/docs/ComboBoxOption/menuPlacement" />

### Section

A combo box can have sections and section headers.

<Example src="ComboBox/docs/ComboBoxOption/section" />

### Footer

A combo box can have a footer.

<Example src="ComboBox/docs/ComboBoxOption/footer" />

### Avatar

A combo box item can contain an avatar.

<Example src="ComboBox/docs/ComboBoxOption/avatar" />

### Count

A combo box item can contain a count using a badge.

<Example src="ComboBox/docs/ComboBoxOption/count" />

### Dynamic Lists

Options and sections can be populated from a hierarchial data structure.
If a section has a header, the `Collection` component can be used to render the child items.

<Example src="ComboBox/docs/ComboBoxOption/dynamicLists" />

### Icons

> A combobox item can contain icons.

Reviewer: ```suggestion
A combo box item can contain icons.
```
